network/gossip/libp2p/discovery.go

```diff
@@ -3,84 +3,80 @@ package libp2p
 import (
 	"context"
 	"fmt"
-	"net"
 	"time"
 
 	"github.com/libp2p/go-libp2p-core/discovery"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/rs/zerolog"
 
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
 	"github.com/onflow/flow-go/network/gossip/libp2p/middleware"
 )
 
 // Discovery implements the discovery.Discovery interface to provide libp2p a way to discover other nodes
 type Discovery struct {
+	ctx     context.Context
 	log     zerolog.Logger
 	overlay middleware.Overlay
 	me      flow.Identifier
-	done    chan struct{}
 }
 
-func NewDiscovery(log zerolog.Logger, overlay middleware.Overlay, me flow.Identifier, done chan struct{}) *Discovery {
-	d := &Discovery{overlay: overlay, log: log, me: me, done: done}
+func NewDiscovery(ctx context.Context, log zerolog.Logger, overlay middleware.Overlay, me flow.Identifier) *Discovery {
+	d := &Discovery{
+		ctx:     ctx,
+		overlay: overlay,
+		log:     log,
+		me:      me,
+	}
 	return d
 }
 
 // Advertise is suppose to advertise this node's interest in a topic to a discovery service. However, we are not using
 // a discovery service hence this function just returns a long duration to reduce the frequency with which libp2p calls it.
-func (d *Discovery) Advertise(_ context.Context, _ string, _ ...discovery.Option) (time.Duration, error) {
-	select {
-	case <-d.done:
-		return 0, fmt.Errorf("middleware stopped")
-	default:
-		return time.Hour, nil
+func (d *Discovery) Advertise(ctx context.Context, _ string, _ ...discovery.Option) (time.Duration, error) {
+	err := ctx.Err()
+	if err != nil {
+		return 0, err
```

Comment:
would be good to append some metadata information on the place where the error happens in both this and subsequent functions: `return 0, fmt.Errorf("could not advertise the node: %w", err)`. Since if such an error happens we are sending a `zero` delay for the next subsequent call. So it would be good to have a clue on it.